src/data/ar-io-data-source.ts

```diff
@@ -0,0 +1,235 @@
+
+  constructor({
+    log,
+    nodeWallet,
+    maxHopsAllowed = DEFAULT_MAX_HOPS_ALLOWED,
+    requestTimeoutMs = DEFAULT_REQUEST_TIMEOUT_MS,
+    updatePeersRefreshIntervalMs = DEFAULT_UPDATE_PEERS_REFRESH_INTERVAL_MS,
+  }: {
+    log: winston.Logger;
+    nodeWallet?: string;
+    maxHopsAllowed?: number;
+    requestTimeoutMs?: number;
+    updatePeersRefreshIntervalMs?: number;
+  }) {
+    this.log = log.child({ class: 'ArIODataSource' });
+    this.nodeWallet = nodeWallet;
+    this.maxHopsAllowed = maxHopsAllowed;
+    this.requestTimeoutMs = requestTimeoutMs;
+    this.updatePeersRefreshIntervalMs = updatePeersRefreshIntervalMs;
+    this.arIO = ArIO.init();
+
+    this.updatePeerList();
+    setInterval(
+      this.updatePeerList.bind(this),
+      this.updatePeersRefreshIntervalMs,
+    );
+  }
+
+  async updatePeerList() {
+    const log = this.log.child({ method: 'updatePeerList' });
+    log.info('Updating peers from ArIO contract');
+
+    try {
+      const response = await this.arIO.getGateways();
+
+      // Convert response to <peerWallet>: <peerUrl> format
+      const peers: Record<string, string> = Object.keys(response).reduce(
+        (acc, key) => {
+          const { protocol, fqdn } = response[key].settings;
+          acc[key] = `${protocol}://${fqdn}`;
+          return acc;
+        },
+        {} as Record<string, string>,
+      );
+
+      // Remove node wallet from peers list if it exists
+      if (this.nodeWallet !== undefined && peers[this.nodeWallet]) {
+        delete peers[this.nodeWallet];
+      }
+
+      this.peers = peers;
+    } catch (error) {
+      log.error('Failed to update peer list', { error });
+      throw new Error('Failed to update peer list');
+    }
+  }
+
+  selectPeer(): string {
+    const log = this.log.child({ method: 'selectPeer' });
+    const keys = Object.keys(this.peers);
+
+    if (keys.length === 0) {
+      log.error('No peers available');
+      throw new Error('No peers available');
+    }
+
+    const randomIndex = randomInt(0, keys.length);
+    return keys[randomIndex];
+  }
+
+  async request({
+    peerAddress,
+    id,
+    headers,
+  }: {
+    peerAddress: string;
+    id: string;
+    headers: { [key: string]: string };
+  }): Promise<AxiosResponse> {
+    const log = this.log.child({ method: 'request' });
+    const path = `/raw/${id}`;
+
+    log.info('Fetching contiguous data from ArIO peer', {
+      id,
+      peerAddress,
+      path,
+    });
+
+    try {
+      const response = await axios.get(`${peerAddress}${path}`, {
+        headers: {
+          'Accept-Encoding': 'identity',
+          ...headers,
+        },
+        responseType: 'stream',
+        timeout: this.requestTimeoutMs,
+      });
+
+      if (response.status !== 200) {
+        throw new Error(`Unexpected status code from peer: ${response.status}`);
+      }
+
+      return response;
+    } catch (error) {
+      log.error('Request to peer failed', { error });
```

Comment:
Let's skip the try/catch here since the caller already handles the error and logs appropriately.